<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Genotype Imputation &mdash; GWASpy 0.1.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Haplotype Phasing" href="phasing.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> GWASpy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="preimp_qc.html">Pre-Imputation QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="pca.html">Principal Component Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="phasing.html">Haplotype Phasing</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Genotype Imputation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#imputation">1. Imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#concat">2. Concat</a></li>
<li class="toctree-l2"><a class="reference internal" href="#arguments-and-options">Arguments and options</a></li>
<li class="toctree-l2"><a class="reference internal" href="#output">Output</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">GWASpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Genotype Imputation</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/imputation.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="genotype-imputation">
<span id="sec-imputation"></span><h1>Genotype Imputation<a class="headerlink" href="#genotype-imputation" title="Permalink to this headline"></a></h1>
<p>Genotype imputation is a process of estimating missing genotypes from the haplotype or genotype reference panel. It
allows you to accurately evaluate the evidence for association at genetic markers that are not directly genotyped.
GWASpy has a module, <code class="code docutils literal notranslate"><span class="pre">imputation</span></code>, for running imputation using IMPUTE5. Because imputation is a computationally
intensive task, we run it on multiple chunks in parallel, then merge the imputed chunks together at the end. This is why
the module is divided into two parts: (1) <code class="code docutils literal notranslate"><span class="pre">impute</span></code>; (2) <code class="code docutils literal notranslate"><span class="pre">concat</span></code>. Below are examples of how to run imputation</p>
<section id="imputation">
<h2>1. Imputation<a class="headerlink" href="#imputation" title="Permalink to this headline"></a></h2>
<ol class="arabic">
<li><p>Command line</p>
<blockquote>
<div><div class="highlight-sh notranslate"><div class="highlight"><pre><span></span>imputation --input-vcfs gs://path/to/vcf_files.txt --samples-file gs://path/to/female_samples.txt --out-dir gs://path/to/output/dir --billing-project project-name --bucket bucket-associated-with-project --run impute --n-samples integer_number_of_samples
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>Python (inside a Python script)</p>
<blockquote>
<div><div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">gwaspy.imputation</span> <span class="k">as</span> <span class="nn">impute</span>
<span class="n">impute</span><span class="o">.</span><span class="n">imputation</span><span class="o">.</span><span class="n">genotype_imputation</span><span class="p">(</span><span class="n">input_vcfs</span> <span class="o">=</span> <span class="s1">&#39;gs://path/to/vcf_files.txt&#39;</span><span class="p">,</span>
          <span class="n">females_file</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">gs</span><span class="p">:</span><span class="o">//</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">female_samples</span><span class="o">.</span><span class="n">txt</span><span class="p">,</span>
          <span class="n">n_samples</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="n">integer_number_of_samples</span><span class="p">,</span>
          <span class="n">n_panel_samples</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">4099</span><span class="p">,</span>
          <span class="n">buffer_region</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">250</span><span class="p">,</span>
          <span class="n">local</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
          <span class="n">billing_project</span> <span class="o">=</span> <span class="s1">&#39;project-name&#39;</span><span class="p">,</span>
          <span class="n">bucket</span> <span class="o">=</span> <span class="s1">&#39;bucket-associated-with-project&#39;</span><span class="p">,</span>
          <span class="n">memory</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;highmem&#39;</span><span class="p">,</span>
          <span class="n">cpu</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">8</span><span class="p">,</span>
          <span class="n">run</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;impute&#39;</span>
          <span class="n">output_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;bcf&#39;</span><span class="p">,</span>
          <span class="n">out_dir</span> <span class="o">=</span> <span class="s1">&#39;gs://path/to/output/dir&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div></blockquote>
</li>
</ol>
</section>
<section id="concat">
<h2>2. Concat<a class="headerlink" href="#concat" title="Permalink to this headline"></a></h2>
<p>After running imputation, we have to merge the imputed chunks together into full chromosomes. Here are examples below
#. Command line</p>
<blockquote>
<div><div class="highlight-sh notranslate"><div class="highlight"><pre><span></span>imputation --input-vcfs gs://path/to/vcf_files.txt --samples-file gs://path/to/female_samples.txt --out-dir gs://path/to/output/dir --billing-project project-name --bucket bucket-associated-with-project --run concat --n-samples integer_number_of_samples
</pre></div>
</div>
</div></blockquote>
<ol class="arabic">
<li><p>Python (inside a Python script)</p>
<blockquote>
<div><div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">gwaspy.imputation</span> <span class="k">as</span> <span class="nn">impute</span>
<span class="n">impute</span><span class="o">.</span><span class="n">imputation</span><span class="o">.</span><span class="n">genotype_imputation</span><span class="p">(</span><span class="n">input_vcfs</span> <span class="o">=</span> <span class="s1">&#39;gs://path/to/vcf_files.txt&#39;</span><span class="p">,</span>
          <span class="n">females_file</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">gs</span><span class="p">:</span><span class="o">//</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">female_samples</span><span class="o">.</span><span class="n">txt</span><span class="p">,</span>
          <span class="n">n_samples</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="n">integer_number_of_samples</span><span class="p">,</span>
          <span class="n">n_panel_samples</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">4099</span><span class="p">,</span>
          <span class="n">buffer_region</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">250</span><span class="p">,</span>
          <span class="n">local</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
          <span class="n">billing_project</span> <span class="o">=</span> <span class="s1">&#39;project-name&#39;</span><span class="p">,</span>
          <span class="n">bucket</span> <span class="o">=</span> <span class="s1">&#39;bucket-associated-with-project&#39;</span><span class="p">,</span>
          <span class="n">memory</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;highmem&#39;</span><span class="p">,</span>
          <span class="n">cpu</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">8</span><span class="p">,</span>
          <span class="n">run</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;concat&#39;</span>
          <span class="n">output_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;bcf&#39;</span><span class="p">,</span>
          <span class="n">out_dir</span> <span class="o">=</span> <span class="s1">&#39;gs://path/to/output/dir&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div></blockquote>
</li>
</ol>
</section>
<section id="arguments-and-options">
<h2>Arguments and options<a class="headerlink" href="#arguments-and-options" title="Permalink to this headline"></a></h2>
<table class="colwidths-given docutils align-default">
<colgroup>
<col style="width: 23%" />
<col style="width: 77%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="code docutils literal notranslate"><span class="pre">--input-vcfs</span></code></p></td>
<td><p>Path to where text file containing VCF(s) for target genotypes paths is</p></td>
</tr>
<tr class="row-odd"><td><p><code class="code docutils literal notranslate"><span class="pre">--samples-file</span></code></p></td>
<td><p>Text file with list of FEMALE samples, one sample ID each line, that are in the dataset. This is crucial for chromosome X imputation as the data is split by sex</p></td>
</tr>
<tr class="row-even"><td><p><code class="code docutils literal notranslate"><span class="pre">--local</span></code></p></td>
<td><p>Type of service. Default is Service backend where jobs are executed on a multi-tenant compute cluster in Google Cloud</p></td>
</tr>
<tr class="row-odd"><td><p><code class="code docutils literal notranslate"><span class="pre">--billing-project</span></code></p></td>
<td><p>Billing project to be used for the job(s)</p></td>
</tr>
<tr class="row-even"><td><p><code class="code docutils literal notranslate"><span class="pre">--bucket</span></code></p></td>
<td><p>Bucket associated with the billing project</p></td>
</tr>
<tr class="row-odd"><td><p><code class="code docutils literal notranslate"><span class="pre">--memory</span></code></p></td>
<td><p>Memory to use for imputation. Options: [<code class="code docutils literal notranslate"><span class="pre">lowmem</span></code>, <code class="code docutils literal notranslate"><span class="pre">standard</span></code>, <code class="code docutils literal notranslate"><span class="pre">highmem</span></code>]. Default is <code class="code docutils literal notranslate"><span class="pre">highmem</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="code docutils literal notranslate"><span class="pre">--cpu-concat</span></code></p></td>
<td><p>CPU to use for the concatenation step. Default is 8</p></td>
</tr>
<tr class="row-odd"><td><p><code class="code docutils literal notranslate"><span class="pre">--n-samples</span></code></p></td>
<td><p>Total number of samples in your dataset. We use this to estimate some of the job resources like storage.</p></td>
</tr>
<tr class="row-even"><td><p><code class="code docutils literal notranslate"><span class="pre">--buffer-region</span></code></p></td>
<td><p>Buffer region to be used during imputation. This helps prevent imputation quality from deteriorating near the edges of the region. Default is 250 KB</p></td>
</tr>
<tr class="row-odd"><td><p><code class="code docutils literal notranslate"><span class="pre">--run</span></code></p></td>
<td><p>Process to run. Options: [<code class="code docutils literal notranslate"><span class="pre">impute</span></code>, <code class="code docutils literal notranslate"><span class="pre">concat</span></code>]. Default is <code class="code docutils literal notranslate"><span class="pre">impute</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="code docutils literal notranslate"><span class="pre">--out-type</span></code></p></td>
<td><p>Output type. Options: [<code class="code docutils literal notranslate"><span class="pre">bcf</span></code>, <code class="code docutils literal notranslate"><span class="pre">vcf</span></code>]. Default is <code class="code docutils literal notranslate"><span class="pre">bcf</span></code> [HIGHLY RECOMMENDED SINCE BCFs ARE GENERALLY MORE EFFICIENT TO WORK WITH AND TAKE UP LESS SPACE]</p></td>
</tr>
<tr class="row-odd"><td><p><code class="code docutils literal notranslate"><span class="pre">--out-dir</span></code></p></td>
<td><p>Path to where output files will be saved</p></td>
</tr>
</tbody>
</table>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline"></a></h2>
<p>The resulting output is a VCF file per chromosome with imputed genotypes.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Concatenating BCFs from imputation by chromosome is slower when the output is VCF compared to a BCF. The size may
also differ significantly between BCF and VCF.</p>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="phasing.html" class="btn btn-neutral float-left" title="Haplotype Phasing" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Martin Lab, Broad Institute.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>