<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Haplotype Phasing &mdash; GWASpy 0.1.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Genotype Imputation" href="imputation.html" />
    <link rel="prev" title="Project PCA (with a reference)" href="pca/project.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> GWASpy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="preimp_qc.html">Pre-Imputation QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="pca.html">Principal Component Analysis</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Haplotype Phasing</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#a-run-scatter-phase-and-concat-in-a-single-command">A. Run scatter, phase, and concat in a single command</a></li>
<li class="toctree-l2"><a class="reference internal" href="#b-run-scatter-phasing-and-concat-steps-in-separate-commands">B. Run scatter, phasing, and concat steps in separate commands.</a></li>
<li class="toctree-l2"><a class="reference internal" href="#c-reference-panels">C. Reference panels</a></li>
<li class="toctree-l2"><a class="reference internal" href="#arguments-and-options">Arguments and options</a></li>
<li class="toctree-l2"><a class="reference internal" href="#output">Output</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="imputation.html">Genotype Imputation</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">GWASpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Haplotype Phasing</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/phasing.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="haplotype-phasing">
<span id="sec-phasing"></span><h1>Haplotype Phasing<a class="headerlink" href="#haplotype-phasing" title="Permalink to this headline"></a></h1>
<p>Knowing the phase of a haplotype can allow us to impute low frequency variants, this makes haplotype phasing an
important step before genotype imputation. GWASpy has a module, <code class="code docutils literal notranslate"><span class="pre">phasing</span></code>, for performing phasing. Phasing can
be run with or without a reference panel using either Eagle2 or SHAPEIT4</p>
<p>In GWASpy, the <code class="code docutils literal notranslate"><span class="pre">phasing</span></code> is divided into 3 parts: (1) <code class="code docutils literal notranslate"><span class="pre">scatter</span></code>; (2) <code class="code docutils literal notranslate"><span class="pre">phase</span></code>; (3) <code class="code docutils literal notranslate"><span class="pre">concat</span></code>. We
first split the input file into multiple smaller chunks with overlapping windows between consecutive windows, run
phasing on each chunk, then concatenate (join) the phased chunks. Running things this way rather than phasing entire
chromosomes speeds up the time it takes to run phasing since we can parallelize the phasing of each chunk. All 3 steps
can be run in a single command (i.e. as a pipeline, DEFAULT &amp; RECOMMENDED) or in separate commands either directly from
the command line or inside a Python script.</p>
<section id="a-run-scatter-phase-and-concat-in-a-single-command">
<h2>A. Run scatter, phase, and concat in a single command<a class="headerlink" href="#a-run-scatter-phase-and-concat-in-a-single-command" title="Permalink to this headline"></a></h2>
<p>You can run the scatter, phase, and concat steps sequentially in a single command. This is the default behaviour in
GWASpy. Below are examples</p>
<ol class="arabic">
<li><p>Command line</p>
<blockquote>
<div><div class="highlight-sh notranslate"><div class="highlight"><pre><span></span>phasing --input-vcf gs://path/to/file.vcf.bgz --out-dir gs://path/to/output/dir --reference GRCh38 --billing-project billing-project
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>Python (inside a Python script)</p>
<blockquote>
<div><div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">gwaspy.phasing</span> <span class="k">as</span> <span class="nn">phase</span>
<span class="n">phase</span><span class="o">.</span><span class="n">phasing</span><span class="o">.</span><span class="n">haplotype_phasing</span><span class="p">(</span><span class="n">input_vcf</span> <span class="o">=</span> <span class="s1">&#39;gs://path/to/file.vcf.bgz&#39;</span><span class="p">,</span>
          <span class="n">vcf_ref</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
          <span class="n">local</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
          <span class="n">billing_project</span> <span class="o">=</span> <span class="s1">&#39;billing-project&#39;</span><span class="p">,</span>
          <span class="n">software</span> <span class="o">=</span> <span class="s1">&#39;shapeit&#39;</span><span class="p">,</span>
          <span class="n">reference</span><span class="o">=</span> <span class="s1">&#39;GRCh38&#39;</span><span class="p">,</span>
          <span class="n">max_win_size_cm</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">10.0</span><span class="p">,</span>
          <span class="n">overlap_size_cm</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">2.0</span><span class="p">,</span>
          <span class="n">scatter_memory</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">26</span><span class="p">,</span>
          <span class="n">cpu</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span>
          <span class="n">threads</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span>
          <span class="n">stages</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;scatter,phase,concat&#39;</span><span class="p">,</span>
          <span class="n">output_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;bcf&#39;</span><span class="p">,</span>
          <span class="n">out_dir</span> <span class="o">=</span> <span class="s1">&#39;gs://path/to/output/dir&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div></blockquote>
</li>
</ol>
</section>
<section id="b-run-scatter-phasing-and-concat-steps-in-separate-commands">
<h2>B. Run scatter, phasing, and concat steps in separate commands.<a class="headerlink" href="#b-run-scatter-phasing-and-concat-steps-in-separate-commands" title="Permalink to this headline"></a></h2>
<p>If for whatever reasons you’d like to run the scatter, phase, and concat steps separately, you can make use of the
<code class="code docutils literal notranslate"><span class="pre">--stages</span></code> (command-line) and <code class="code docutils literal notranslate"><span class="pre">stages</span></code> (Python script) arguments to specify which stage
[<code class="code docutils literal notranslate"><span class="pre">scatter</span></code>, <code class="code docutils literal notranslate"><span class="pre">phase</span></code>, <code class="code docutils literal notranslate"><span class="pre">concat</span></code>] you want to run. It’s important to note that even though you can run things this way, phase is dependent on results from scatter and concat on results
from phase i.e. you cannot run phasing without having ran scatter prior.</p>
</section>
<section id="c-reference-panels">
<h2>C. Reference panels<a class="headerlink" href="#c-reference-panels" title="Permalink to this headline"></a></h2>
<p>In some cases, including a reference panel when phasing might improve accuracy. By default, GWASpy runs phasing without
a reference panel. If the user wants to use a reference panel, there are two options</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If the data set you wish to phase contains more than twice as many samples as the largest reference panel
available to you, then using a reference panel is unlikely to give much of a boost in phasing accuracy.</p>
</div>
<p><strong>C1.  HGDP+1KG dataset</strong></p>
<div class="highlight-sh notranslate"><div class="highlight"><pre><span></span>phasing --input-vcf gs://path/to/file.vcf.bgz --out-dir gs://path/to/output/dir --reference GRCh38 --billing-project billing-project --vcf-ref hgdp_1kg
</pre></div>
</div>
<p><strong>C2. Their own reference panel</strong></p>
<p>Say you have your reference panel files by chromosomes stored in gs://ref_panel/ALL.chr{1..22,X}.vcf,
you would pass the path to <code class="code docutils literal notranslate"><span class="pre">--vcf-ref</span></code> as gs://ref_panel/ALL.chr<strong>CNUMBER</strong>.vcf,
GWASpy uses <strong>CNUMBER</strong> as a placeholder for the chromosomes. Then you can run phasing as:</p>
<div class="highlight-sh notranslate"><div class="highlight"><pre><span></span>phasing --input-vcf gs://path/to/file.vcf.bgz --out-dir gs://path/to/output/dir --reference GRCh38 --billing-project project-name --vcf-ref gs://ref_panel/ALL.chrCNUMBER.vcf
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ol class="arabic simple">
<li><p>If you’re using your own reference panel, make sure the files are bgzip compressed.</p></li>
<li><p>Chromosome X reference file must be name X and not 23</p></li>
</ol>
</div>
</section>
<section id="arguments-and-options">
<h2>Arguments and options<a class="headerlink" href="#arguments-and-options" title="Permalink to this headline"></a></h2>
<table class="colwidths-given docutils align-default">
<colgroup>
<col style="width: 23%" />
<col style="width: 77%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="code docutils literal notranslate"><span class="pre">--input-vcf</span></code></p></td>
<td><p>Path to where VCF file to be phased is</p></td>
</tr>
<tr class="row-odd"><td><p><code class="code docutils literal notranslate"><span class="pre">--vcf-ref</span></code></p></td>
<td><p>VCF file for reference haplotypes if phasing with a reference panel</p></td>
</tr>
<tr class="row-even"><td><p><code class="code docutils literal notranslate"><span class="pre">--local</span></code></p></td>
<td><p>Type of service. Default is Service backend where jobs are executed on a multi-tenant compute cluster in Google Cloud</p></td>
</tr>
<tr class="row-odd"><td><p><code class="code docutils literal notranslate"><span class="pre">--billing-project</span></code></p></td>
<td><p>Billing project to be used for the job(s)</p></td>
</tr>
<tr class="row-even"><td><p><code class="code docutils literal notranslate"><span class="pre">--software</span></code></p></td>
<td><p>Software to use for phasing. Options: [<code class="code docutils literal notranslate"><span class="pre">eagle</span></code>, <code class="code docutils literal notranslate"><span class="pre">shapeit</span></code>]. Default is <code class="code docutils literal notranslate"><span class="pre">eagle</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="code docutils literal notranslate"><span class="pre">--reference</span></code></p></td>
<td><p>Genome reference build. Default is GRCh38. Options: [<code class="code docutils literal notranslate"><span class="pre">GRCh37</span></code>, <code class="code docutils literal notranslate"><span class="pre">GRCh38</span></code>]</p></td>
</tr>
<tr class="row-even"><td><p><code class="code docutils literal notranslate"><span class="pre">--max-win-size-cm</span></code></p></td>
<td><p>Maximum window size to use when chunking the input file. Default is 10.0</p></td>
</tr>
<tr class="row-odd"><td><p><code class="code docutils literal notranslate"><span class="pre">--overlap-size-cm</span></code></p></td>
<td><p>Size of overlap between consecutive overlapping windows. Default is 2.0</p></td>
</tr>
<tr class="row-even"><td><p><code class="code docutils literal notranslate"><span class="pre">--cpu</span></code></p></td>
<td><p>Number of CPUs to use in phasing. Default is 4. [TO BE CHANGED]</p></td>
</tr>
<tr class="row-odd"><td><p><code class="code docutils literal notranslate"><span class="pre">--scatter-mem</span></code></p></td>
<td><p>Memory to use for scattering input into chunks before phasing. [TO BE CHANGED]</p></td>
</tr>
<tr class="row-even"><td><p><code class="code docutils literal notranslate"><span class="pre">--threads</span></code></p></td>
<td><p>Number of threads to use in phasing. Default is 3. [TO BE CHANGED]</p></td>
</tr>
<tr class="row-odd"><td><p><code class="code docutils literal notranslate"><span class="pre">--stages</span></code></p></td>
<td><p>Process(es) to run. Default is <code class="code docutils literal notranslate"><span class="pre">scatter,phase,concat</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="code docutils literal notranslate"><span class="pre">--out-type</span></code></p></td>
<td><p>Output type. Options: [<code class="code docutils literal notranslate"><span class="pre">bcf</span></code>, <code class="code docutils literal notranslate"><span class="pre">vcf</span></code>]. Default is <code class="code docutils literal notranslate"><span class="pre">bcf</span></code> [HIGHLY RECOMMENDED SINCE BCFs ARE GENERALLY FASTER TO WORK WITH AND TAKE UP LESS SPACE]</p></td>
</tr>
<tr class="row-odd"><td><p><code class="code docutils literal notranslate"><span class="pre">--out-dir</span></code></p></td>
<td><p>Path to where output files will be saved</p></td>
</tr>
</tbody>
</table>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline"></a></h2>
<p>For both Eagle and SHAPEIT, the resulting output is a VCF file per chromosome with phased haplotypes.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>By default, Eagle will output a VCF file with phased GT and other fields that were in the unphased VCF, whereas
SHAPEIT will ONLY output the GT field. This will result in phased files generated using Eagle being bigger in size
than those generate using SHAPEIT.</p>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="pca/project.html" class="btn btn-neutral float-left" title="Project PCA (with a reference)" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="imputation.html" class="btn btn-neutral float-right" title="Genotype Imputation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Martin Lab, Broad Institute.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>